### PR TITLE
Add linker annotations for System.Text.Json

### LINKED ISSUE #34449: System.Text.Json.Serialization.JsonStringEnumConverter isn't linker-safe
Most other converters in S.T.J. (as far as I know) are safe because their default constructors are referenced statically. However, if you use `JsonStringEnumConverter` like this:

```cs
[JsonConverter(typeof(JsonStringEnumConverter))]
public enum MyEnum { A, B, C }
```

then there is no static reference to `JsonStringEnumConverter`'s default constructor, so the linker will strip it out. Then at runtime it will fail when `JsonSerializerOptions.GetConverterFromAttribute` tries to call `Activator.CreateInstance(converterType)` (where `converterType = typeof(JsonStringEnumConverter)`). This was originally reported at https://github.com/dotnet/aspnetcore/issues/19086

For Blazor 3.2.0, we'll put in a workaround into Blazor's built-in linker config. However it would make sense to fix the underlying issue in S.T.J. in the longer term.

### Suggested fix

Either `System.Text.Json.dll` should embed its own linker config file preserving the constructors for all converter types, or consider extending the set of types listed in `JsonSerializerOptions.DefaultSimpleConverters` to include an instantiation of `JsonStringEnumConverter` (or have any other static reference to its constructor).

This applies to all converter types that may be used at runtime. The only one I know of that's causing a problem is `JsonStringEnumConverter`, but I didn't exhaustively check that all the others are linker-safe too.

### PR DESCRIPTION
**Pull request by layomia**
- Fixes https://github.com/dotnet/runtime/issues/34449 (System.Text.Json.Serialization.JsonStringEnumConverter isn't linker-safe)
- Adds annotations to make simple serializer usages linker safe (e.g. deserializing most collections and POCOs with primitive members). More complex object graphs e.g. POCOs with POCO properties will probably need [`DynamicallyAccessedMemberTypes.RecursiveProperties/Constructors`](https://github.com/mono/linker/issues/1087).
    - @steveharter mentions another view point/approach in https://github.com/mono/linker/issues/1087#issuecomment-615479254.

Deserializing immutable collections is not linker safe - will address with https://github.com/dotnet/runtime/issues/38593.

### CODE REVIEW

File: src/libraries/System.Text.Json/tests/TrimmingTests/EnumConverterTest.cs
Comment (MichalStrehovsky) [inline src/libraries/System.Text.Json/tests/TrimmingTests/EnumConverterTest.cs:26]: Might be better not to touch the properties and leave them at their default value (this applies to all the other tests too) - linker can potentially have heuristics that say: if a setter is kept, also keep the getter.

Either keep the properties at their default values, or instead of auto properties, implement this with an explicit private backing field and only ever touch the backing field with C# - don't reference the properties from C# at all.

Comment (MichalStrehovsky) [reply src/libraries/System.Text.Json/tests/TrimmingTests/EnumConverterTest.cs:26]: (Basically, we need to make the properties as unreferenced as possible so that the linker is as free to strip it and the only reason why it didn't strip it are the annotations.)

File: src/libraries/System.Text.Json/tests/TrimmingTests/SerializerEntryPoint/DeserializeAsync.FromStream.TypedObject.cs
Comment (eerhardt) [inline src/libraries/System.Text.Json/tests/TrimmingTests/SerializerEntryPoint/DeserializeAsync.FromStream.TypedObject.cs:30]: Can you use a `static async Task<int>Main` method instead? Then you can just await as normal without having to do a Task.Run.

File: src/libraries/System.Text.Json/tests/TrimmingTests/System.Text.Json.TrimmingTests.proj
Comment (eerhardt) [inline src/libraries/System.Text.Json/tests/TrimmingTests/System.Text.Json.TrimmingTests.proj:4]: I think this drops the `StackOrQueueNotRootedTests.cs`. You either need to have all files explicitly included, or none.

@joperezr - maybe we should re-think how we glob the test files so this kind of mistake doesn't happen. What if we used a convention that each test case ended with `Test.cs` (or a exclude pattern like `.NotTest.cs`), and we only included test files by default? Then when a test wanted to include an extra file, it could use the MSBuild `Update` syntax:

```xml
    <TestConsoleAppSourceFiles Update="SerializerEntryPoint\Deserialize.FromReader.BoxedObject.cs">
      <AdditionalSourceFiles>Helper.cs</AdditionalSourceFiles>
    </TestConsoleAppSourceFiles>
```

Thoughts?

Comment (joperezr) [reply src/libraries/System.Text.Json/tests/TrimmingTests/System.Text.Json.TrimmingTests.proj:4]: That sounds good to me. Honestly the original thought was to move to explicitly adding them to the projects and stop doing globbing in order to allow customization and also to mimic what we do in the rest of the repo where we by default always disable glob adding of source files, but I decided to support both modes in order to not conflict my PR with other PRs from you and @layomia which were open at the same time. All that said, if we want to continue the support for glob adding tests, your suggestion sounds great to me, or if we instead want to mimic the rest of the repo and do explicit includes only, I'm fine with that too, both are very trivial changes.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Span.cs
Comment (joperezr) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Span.cs:27]: @eerhardt I believe that I've asked this already but is there a requirement for us to also update the ref code with these attributes? I know that they only used on the linker fwith the actual implementation, but I'm not sure if we should add them to the refs so that other projects that depend on System.Text.Json throw linker warnings saying they have to add or propagate these attributes as well.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Helpers.cs
Comment (joperezr) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Helpers.cs:14]: Don't we need the constructor as well on write?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonStringEnumConverter.cs
Comment (joperezr) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonStringEnumConverter.cs:57]: Why was this removed? Is the linker recognizing these dependencies now?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableConverterFactory.cs
Comment (joperezr) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableConverterFactory.cs:40]: Are these not required anymore due to the other ones you are adding?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableConverterFactoryHelpers.cs
Comment (joperezr) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableConverterFactoryHelpers.cs:142]: small NIT: We usually try to avoid checking in TODO comments on the code, any reason why we can' get to these two on the same PR? Are we blocked by a linker feature?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Span.cs
Comment (eerhardt) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Span.cs:27]: Good catch. Yes, we need to add these in the ref assemblies so when we link during the dotnet/runtime build the linker can see the attributes.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Helpers.cs
Comment (eerhardt) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Helpers.cs:14]: No - we don't construct new objects during write. (Note that this is writing the objects to JSON)

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Span.cs
Comment (joperezr) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Span.cs:27]: What about the other PRs where we've added attributes? Y suppose DynamicDependency ones don't really need to be propagated to the refs, but I believe we have added DynamicAccessedMembers attributes on other PRs as well.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableConverterFactoryHelpers.cs
Comment (danmosemsft) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableConverterFactoryHelpers.cs:142]: Or just omit the TODO, that's what we have issues for.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Span.cs
Comment (eerhardt) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Span.cs:27]: > DynamicDependency ones don't really need to be propagated to the refs

Correct. Those are explicitly left out.

https://github.com/dotnet/runtime/blob/ecdb7f36e0d97e204194ee23a972160471b44622/eng/DefaultGenApiDocIds.txt#L10

> but I believe we have added DynamicAccessedMembers attributes on other PRs as well.

The ones that @MichalStrehovsky opened have updated the refs. e.g. https://github.com/dotnet/runtime/pull/38331 and https://github.com/dotnet/runtime/pull/38021.

We should update the `ref` assemblies on any place we've added but haven't updated the `ref` files.

Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Span.cs:27]: Fixed, thanks.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonStringEnumConverter.cs
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonStringEnumConverter.cs:57]: Yes - this situation is similar to the discussion in https://github.com/dotnet/runtime/pull/38216#discussion_r445670381/ Because there are `typeof` usages in the code, the linker can recognize these dependencies.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Span.cs
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Span.cs:27]: > We should update the ref assemblies on any place we've added but haven't updated the ref files.

I'll scan through and make a PR to added them, if any.

File: src/libraries/System.Text.Json/tests/TrimmingTests/System.Text.Json.TrimmingTests.proj
Comment (layomia) [reply src/libraries/System.Text.Json/tests/TrimmingTests/System.Text.Json.TrimmingTests.proj:4]: I'm in favor of requiring that the files be explicitly added, so that we are more consistent with the default behavior in the rest of the repo. I'll make a PR for this if there's no opposition to this.

Comment (eerhardt) [reply src/libraries/System.Text.Json/tests/TrimmingTests/System.Text.Json.TrimmingTests.proj:4]: sounds good to me

File: src/libraries/System.Text.Json/ref/System.Text.Json.csproj
Comment (joperezr) [inline src/libraries/System.Text.Json/ref/System.Text.Json.csproj:20]: This itemgroup is wrong, you should never compile implementation code (specially from corelib) via a reference project. Instead you should be referencing the contract where we are exposing DynamicallyAccessedMembersAttribute to get that reference

Comment (joperezr) [reply src/libraries/System.Text.Json/ref/System.Text.Json.csproj:20]: For the other frameworks where this type doesn't exist, you should instead just #ifdef or separate that into a different file that gets conditionally includded only for netcoreappcurrent.

Comment (layomia) [reply src/libraries/System.Text.Json/ref/System.Text.Json.csproj:23]: Fixed, thanks a lot!

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/DynamicallyAccessedMembersAttribute.cs
Comment (joperezr) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/DynamicallyAccessedMembersAttribute.cs:29]: I don't think I understand why you need this?

File: src/libraries/System.Text.Json/ref/System.Text.Json.csproj
Comment (joperezr) [inline src/libraries/System.Text.Json/ref/System.Text.Json.csproj:21]: NIT: I would change the name to just be `System.Text.Json.netcoreapp.cs` and the other one to be `System.Text.Json.non-netcoreapp.cs`

File: src/libraries/System.Text.Json/src/System.Text.Json.csproj
Comment (joperezr) [inline src/libraries/System.Text.Json/src/System.Text.Json.csproj:241]: @eerhardt what do we expect to do here for configurations that are not net5.0? Is the plan to compile them like he is doing here or should we instead have one implementation with the attributes and one without, or should we ifdef the code?

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/DynamicallyAccessedMembersAttribute.cs
Comment (joperezr) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/DynamicallyAccessedMembersAttribute.cs:29]: I understand why you need it now. I'm not sure I agree with what we are doing here to be honest, but I'm fine if that is the direction we are going to take.

File: src/libraries/System.Text.Json/src/System.Text.Json.csproj
Comment (layomia) [reply src/libraries/System.Text.Json/src/System.Text.Json.csproj:241]: FWIW - having two implementations would add the maintenance cost of making sure they are always in sync. ifdef-ing is a bit better but IMO, the current approach is the cleanest/least invasive from a library author/maintainer point of view.

Comment (eerhardt) [reply src/libraries/System.Text.Json/src/System.Text.Json.csproj:241]: > what do we expect to do here for configurations that are not net5.0?

This was discussed here: https://github.com/dotnet/runtime/issues/36656#issuecomment-630345623

For this case, I personally would just `#if` around the attributes in the existing `System.Text.Json.cs` file, just like we do here:

https://github.com/dotnet/runtime/blob/9de1a7e044b70f509c3a974ec717a94486e5391b/src/libraries/System.Drawing.Common/ref/System.Drawing.Common.cs#L671-L674

That way we aren't duplicating everything. Note there will be more annotations in System.Text.Json in the future, for example `JsonConverterAttribute` and `JsonConverter<T>`.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/DynamicallyAccessedMembersAttribute.cs
Comment (eerhardt) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/DynamicallyAccessedMembersAttribute.cs:29]: Yes, that is the direction we've decided to take. See https://github.com/dotnet/runtime/issues/36656#issuecomment-630345623

The other attributes in this folder already have it:

https://github.com/dotnet/runtime/blob/8126962457be677cb8b9313d5d9f26849da77df2/src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/DynamicDependencyAttribute.cs#L17-L22

https://github.com/dotnet/runtime/blob/8126962457be677cb8b9313d5d9f26849da77df2/src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/NullableAttributes.cs#L9-L14

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableConverterFactoryHelpers.cs
Comment (eerhardt) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableConverterFactoryHelpers.cs:139]: (nit) can we use the above `const string` fields for these strings? I think it would be easier to track. Same for the assembly name.

File: src/libraries/System.Text.Json/tests/TrimmingTests/CollectionConverters/CollectionConvertersTest.IImmutableDictionary.cs
Comment (eerhardt) [inline src/libraries/System.Text.Json/tests/TrimmingTests/CollectionConverters/CollectionConvertersTest.IImmutableDictionary.cs:11]: (nit) why are all the Immutable* types split into their own tests, but all the other collections are all slammed together into 1 test?

File: src/libraries/System.Text.Json/tests/TrimmingTests/Helper.cs
Comment (eerhardt) [inline src/libraries/System.Text.Json/tests/TrimmingTests/Helper.cs:100]: This may be a bit of a stretch, but are we concerned that when we are testing `Deserialize` we are also calling `Serialize`? Could there ever be a bug where if the app ONLY called `Deserialize`, something was broken? But we are always doing both in our tests.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Attributes/JsonConverterAttribute.cs
Comment (eerhardt) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Attributes/JsonConverterAttribute.cs:26]: These don't seem to have corresponding entries in the `ref` file. Or am I missing it?

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/DynamicallyAccessedMembersAttribute.cs
Comment (joperezr) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/DynamicallyAccessedMembersAttribute.cs:29]: I see, honestly I don't fully agree with this approach (as opposed to having them available down level) since we probably want to allow library developers that target netstandard or < .NET5.0 to be able to annotate their libraries as well, specially if these will be libraries used for blazor wasm. Given the attribute implementation is fully netstandard2.0 compatible, I would opt instead for having compatibility packs (similar to what we did with AsyncInterfaces) for shipping the attributes down level so that we a) don't have to compile these attributes internally on a bunch of libraries and b) allow third party library writters to consume them. I don't want to push to much on doing this but just wanted to share my opinion 😄 cc: @stephentoub

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Attributes/JsonConverterAttribute.cs
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Attributes/JsonConverterAttribute.cs:26]: Thanks. I've added them now.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableConverterFactoryHelpers.cs
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableConverterFactoryHelpers.cs]: I've addressed the TODO in this PR. I was initially considering an alternative approach which might not have required the added `DynamicDependencyAttribute` usages, but I can explore this in the future since avoiding the attributes is not a priority right now.

File: src/libraries/System.Text.Json/tests/TrimmingTests/CollectionConverters/CollectionConvertersTest.IImmutableDictionary.cs
Comment (layomia) [reply src/libraries/System.Text.Json/tests/TrimmingTests/CollectionConverters/CollectionConvertersTest.IImmutableDictionary.cs:11]: I've split all the collection types into their own tests.

File: src/libraries/System.Text.Json/src/System.Text.Json.csproj
Comment (layomia) [reply src/libraries/System.Text.Json/src/System.Text.Json.csproj:241]: I've updated the ref file to use `#ifdef`s.

File: src/libraries/System.Text.Json/tests/TrimmingTests/Helper.cs
Comment (layomia) [reply src/libraries/System.Text.Json/tests/TrimmingTests/Helper.cs:100]: The intention of these collection tests is to exercise the warm-up code paths of the serializer which is the same for both serialization and deserialization, so we can test either side. Here, we are checking that the reflection-based dependencies like the collection converters, `CreateRange` methods for immutable collections, `Push`/`Enqueue` for `Stack`/`Queue` etc., are rooted when needed.

The tests for each serializer method overload are found in the `TrimmingTest\SerializerEntryPoint` directory. For each of these, we test a POCO and collection. Issues with serialization or deserialization specifically would be pointed out by these tests.

File: src/libraries/System.Text.Json/ref/System.Text.Json.csproj
Comment (eerhardt) [inline src/libraries/System.Text.Json/ref/System.Text.Json.csproj:8]: You don't need to do this yourself. The SDK automatically creates define constants for each TFM. And it also creates a "version-less" define as well.

So instead, you can delete these lines, and in the .cs file you can say: `#if NETCOREAPP`

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableConverterFactoryHelpers.cs
Comment (eerhardt) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableConverterFactoryHelpers.cs:145]: `ImmutableCollectionsNamespace` -> The attribute takes an assembly name, not a namespace. It would be more understandable to call this `ImmutableCollectionsAssembly`.

File: src/libraries/System.Text.Json/tests/TrimmingTests/SerializerEntryPoint/Deserialize.FromReader.BoxedObject.cs
Comment (eerhardt) [inline src/libraries/System.Text.Json/tests/TrimmingTests/SerializerEntryPoint/Deserialize.FromReader.BoxedObject.cs:23]: I think we should verify the object is correct.
```suggestion
            if (arr == null || arr.Length != 1 || arr[0] != 1)
```

**Review by eerhardt — APPROVED** [review]
LGTM

**Comment by safern** [comment]
Merging per @layomia's request.

**Pull request merged**
